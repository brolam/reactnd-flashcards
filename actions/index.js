export const RECEIVE_DECKS = 'RECEIVE_DECKS'
export const SELECT_DECK = 'SELECT_DECK'
export const SET_APP_STATE = 'SET_APP_STATE'
export const SELECT_QUIZ = 'SELECT_QUIZ'
export const APP_STATES = {
  NAVIGATE_DECK: 'navigate_deck',
  NEW_DECK: 'newDeck',
  STARTED_QUIZ: 'started_quiz'
}

export function receiveDecks(decks) {
  return {
    type: RECEIVE_DECKS,
    decks,
  }
}

export function selectDeck(deckKey, deckQuizzes) {
  return {
    type: SELECT_DECK,
    deckKey,
    deckQuizzes,
  }
}

export function setAppState(appState) {
  return {
    type: SET_APP_STATE,
    appState
  }
}

export function selectQuiz(index) {
  return {
    type: SELECT_QUIZ,
    selectIndexQuiz: index
  }
}