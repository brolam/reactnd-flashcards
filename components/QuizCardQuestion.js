import React from 'react'
import { View, Text, TouchableOpacity, Button, StyleSheet } from 'react-native'
import PropTypes from 'prop-types'
import cardStyle from '../styles/cardStyles'
import buttonStyle from '../styles/buttonStyles'
import { green, red, white, } from '../styles/colors';
import { showQuizAnswer } from '../actions/index';
import { FontAwesome } from '@expo/vector-icons'

export default function QuizCardQuestion({
  deck,
  selectedIndexQuiz = 0,
  onAnswer,
  showQuizAnswer = false,
  dispatch = (action) => { console.log('dispatch') },
  onEditQuiz = () => { console.log('Edit Card') } }) {
  const quiz = deck.quizzes[selectedIndexQuiz]
  return (
    <View style={[cardStyle, styles.container]}>
      <View style={styles.headerContainer}>
        <Text style={styles.headerContainerItemLeft} >
          {selectedIndexQuiz + 1}/{deck.amountOfCards}
        </Text>
        <TouchableOpacity style={styles.headerContainerItemRight}
          onPress={() => onEditQuiz()}>
          <FontAwesome style={styles.buttonEditCard} name='edit' />
        </TouchableOpacity>
      </View>
      {showQuizAnswer
        ? <Answer quiz={quiz} dispatch={dispatch} />
        : <Question quiz={quiz} dispatch={dispatch} />
      }
      <View style={styles.buttonsContainer}>
        <TouchableOpacity
          style={[buttonStyle, styles.buttonAnswerCorrect]}
          onPress={() => onAnswer(deck, selectedIndexQuiz, true)}>
          <Text style={styles.textButtons}>Correct?</Text>
        </TouchableOpacity>
        <TouchableOpacity
          style={[buttonStyle, styles.buttonAnswerIncorrect]}
          onPress={() => onAnswer(deck, selectedIndexQuiz, false)}>
          <Text style={styles.textButtons}>Incorrect?</Text>
        </TouchableOpacity>
      </View>
    </View>

  )
}

function Question(props) {
  const { quiz, dispatch } = props
  return (
    <View>
      <Text style={styles.question}>{quiz.question}</Text>
      <TouchableOpacity
        onPress={() => dispatch(showQuizAnswer(true))}>
        <Text style={styles.buttonShowAnswer}>show answer</Text>
      </TouchableOpacity>
    </View>
  )
}

function Answer(props) {
  const { quiz, dispatch } = props
  return (
    <View>
      <Text style={styles.answer}>{quiz.answer}</Text>
      <TouchableOpacity
        onPress={() => dispatch(showQuizAnswer(false))}>
        <Text style={styles.buttonShowAnswer}>show question</Text>
      </TouchableOpacity>
    </View>
  )
}

QuizCardQuestion.propTypes = {
  deck: PropTypes.shape({
    amountOfCards: PropTypes.number.isRequired,
    quizzes: PropTypes.array.isRequired,
  }),
  selectedIndexQuiz: PropTypes.number.isRequired
}

const styles = StyleSheet.create({
  container: {
    flexDirection: 'column',
    justifyContent: 'space-around',
    alignItems: 'center',
    flex: 1,
    marginBottom: 10,
    padding: 10,
  },
  headerContainer: {
    flexDirection: 'row',
    margin: 5,
    marginTop: 18,
    position: 'absolute',
    top: 0
  },
  headerContainerItemLeft: {
    flex: 1,
  },
  headerContainerItemRight: {
    flex: 1,
    alignItems: 'flex-end',
  },
  question: {
    fontSize: 32,
    fontWeight: '400',
    textAlign: 'center',
  },
  answer: {
    fontSize: 36,
    fontWeight: '500',
    textAlign: 'center',
  },
  buttonShowAnswer: {
    color: 'blue',
    fontSize: 16,
    textAlign: 'center',
    paddingTop: 8,
    paddingBottom: 8,
    marginBottom: 30,
  },
  buttonsContainer: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    flexWrap: 'wrap',
    marginTop: 30,
    marginBottom: 20,
    position: 'absolute',
    bottom: 0
  },
  buttonAnswerCorrect: {
    backgroundColor: green,
    marginRight: 5
  },
  buttonAnswerIncorrect: {
    backgroundColor: red,
    marginLeft: 5
  },
  textButtons: {
    color: white,
    fontSize: 18,
    fontWeight: '600',
  },
  buttonEditCard: {
    fontSize: 18,
  },
});