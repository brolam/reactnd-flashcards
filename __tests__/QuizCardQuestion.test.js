import React from 'react'
import QuizCardQuestion from '../components/QuizCardQuestion'
import { showQuizAnswer } from '../actions/index';

const quizzesDummy = [
  { question: 'One Quetion', answer:'One Answer' },
  { question: 'Two Quetion', answer:'Two Answer' },
  { question: 'Three Quetion', answer:'Three Answer' }
]
const deckDummy = { key: 'one-item', title: 'One Deck', amountOfCards: 3, quizzes: quizzesDummy }

test('renders without crashing', () => {
  const wrapper = shallow(<QuizCardQuestion
    deck={deckDummy}
    quizzes={quizzesDummy}
    selectedIndexQuiz={0} />);
  expect(wrapper).toMatchSnapshot();
});

test('show quiz answer', () => {
  const wrapper = shallow(<QuizCardQuestion
    deck={deckDummy}
    quizzes={quizzesDummy}
    selectedIndexQuiz={0}
    showQuizAnswer={true} />);
  expect(wrapper).toMatchSnapshot();
});

test('dispatch show quiz answer', () => {
  const spyDispatch = jest.fn()
  const quizCardQuestion = mount(<QuizCardQuestion
    deck={deckDummy}
    quizzes={quizzesDummy}
    selectedIndexQuiz={0}
    dispatch={spyDispatch} />);
    const buttonShowAnswer = quizCardQuestion.find('TouchableOpacity').at(0)
    buttonShowAnswer.props().onPress()
    expect(spyDispatch).toHaveBeenCalledWith(showQuizAnswer(true))
});

test('dispatch show quiz question', () => {
  const spyDispatch = jest.fn()
  const quizCardQuestion = mount(<QuizCardQuestion
    deck={deckDummy}
    quizzes={quizzesDummy}
    selectedIndexQuiz={0}
    showQuizAnswer={true}
    dispatch={spyDispatch} />);
    const buttonShowQuestion = quizCardQuestion.find('TouchableOpacity').at(0)
    buttonShowQuestion.props().onPress()
    expect(spyDispatch).toHaveBeenCalledWith(showQuizAnswer(false))
});
