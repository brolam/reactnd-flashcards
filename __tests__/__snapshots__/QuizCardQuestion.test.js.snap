// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<View
  style={
    Array [
      Array [
        Object {
          "backgroundColor": "#fff",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginLeft": 10,
          "marginRight": 10,
          "marginTop": 8,
          "padding": 20,
          "shadowColor": "rgba(0, 0, 0, 0.24)",
          "shadowOpacity": 0.8,
        },
        Object {
          "borderRadius": 8,
          "shadowOffset": Object {
            "height": 3,
            "width": 0,
          },
          "shadowRadius": 3,
        },
      ],
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "space-around",
        "marginBottom": 10,
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignSelf": "stretch",
        "flexDirection": "column",
        "justifyContent": "space-around",
        "marginLeft": 8,
        "marginTop": 8,
        "position": "absolute",
        "top": 0,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
    >
      1
      /
      3
    </Text>
  </View>
  <Question
    dispatch={[Function]}
    quiz={
      Object {
        "answer": "One Answer",
        "question": "One Quetion",
      }
    }
  />
  <View
    style={
      Object {
        "bottom": 0,
        "flexDirection": "row",
        "flexWrap": "wrap",
        "justifyContent": "space-around",
        "marginBottom": 20,
        "marginTop": 30,
        "position": "absolute",
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
      style={
        Array [
          Array [
            Object {
              "alignItems": "center",
            },
            Object {
              "backgroundColor": "#fff",
              "borderColor": "#292477",
              "borderRadius": 3,
              "borderWidth": 1,
              "padding": 5,
              "paddingLeft": 25,
              "paddingRight": 25,
            },
          ],
          Object {
            "backgroundColor": "#3cb371",
            "margin": 5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 24,
          }
        }
      >
        Correct?
      </Text>
    </TouchableOpacity>
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
      style={
        Array [
          Array [
            Object {
              "alignItems": "center",
            },
            Object {
              "backgroundColor": "#fff",
              "borderColor": "#292477",
              "borderRadius": 3,
              "borderWidth": 1,
              "padding": 5,
              "paddingLeft": 25,
              "paddingRight": 25,
            },
          ],
          Object {
            "backgroundColor": "#ff0000",
            "margin": 5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 24,
          }
        }
      >
        Incorrect?
      </Text>
    </TouchableOpacity>
  </View>
</View>
`;

exports[`show quiz answer 1`] = `
<View
  style={
    Array [
      Array [
        Object {
          "backgroundColor": "#fff",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginLeft": 10,
          "marginRight": 10,
          "marginTop": 8,
          "padding": 20,
          "shadowColor": "rgba(0, 0, 0, 0.24)",
          "shadowOpacity": 0.8,
        },
        Object {
          "borderRadius": 8,
          "shadowOffset": Object {
            "height": 3,
            "width": 0,
          },
          "shadowRadius": 3,
        },
      ],
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "space-around",
        "marginBottom": 10,
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignSelf": "stretch",
        "flexDirection": "column",
        "justifyContent": "space-around",
        "marginLeft": 8,
        "marginTop": 8,
        "position": "absolute",
        "top": 0,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
    >
      1
      /
      3
    </Text>
  </View>
  <Answer
    dispatch={[Function]}
    quiz={
      Object {
        "answer": "One Answer",
        "question": "One Quetion",
      }
    }
  />
  <View
    style={
      Object {
        "bottom": 0,
        "flexDirection": "row",
        "flexWrap": "wrap",
        "justifyContent": "space-around",
        "marginBottom": 20,
        "marginTop": 30,
        "position": "absolute",
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
      style={
        Array [
          Array [
            Object {
              "alignItems": "center",
            },
            Object {
              "backgroundColor": "#fff",
              "borderColor": "#292477",
              "borderRadius": 3,
              "borderWidth": 1,
              "padding": 5,
              "paddingLeft": 25,
              "paddingRight": 25,
            },
          ],
          Object {
            "backgroundColor": "#3cb371",
            "margin": 5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 24,
          }
        }
      >
        Correct?
      </Text>
    </TouchableOpacity>
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
      style={
        Array [
          Array [
            Object {
              "alignItems": "center",
            },
            Object {
              "backgroundColor": "#fff",
              "borderColor": "#292477",
              "borderRadius": 3,
              "borderWidth": 1,
              "padding": 5,
              "paddingLeft": 25,
              "paddingRight": 25,
            },
          ],
          Object {
            "backgroundColor": "#ff0000",
            "margin": 5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 24,
          }
        }
      >
        Incorrect?
      </Text>
    </TouchableOpacity>
  </View>
</View>
`;
